### DISCUSSION TITLE
Adding device support (88ddrc)

### DISCUSSION THREAD

**Discussion by jvdh1978** [mrene/minidsp-rs#213] [Q&A]
I have been monitoring and communicating with my ddrc-88d. I can work out the data sent back and forth between the MiniDSP app and the device with wireshark. Do you have some document to add full device support to you app? The instruction in your readme is clearly marked as a TODO ;-).

**Comment by mrene** [comment]
Oh hi!

There are a few steps, the first one is to add the xml configuration file from the plugin app, and turn it into a device file through the `codegen` subcommand (it contains addresses and their related names).

You can take example from the [2x4hd](https://github.com/mrene/minidsp-rs/tree/54e6200d8bfbc0c62ad0f6d01c0d23bcbb3de809/devtools/src/codegen/m2x4hd). Once you have a basic setup (you don't have to have any values in the device to get started, you'll do this next), add it to codegen_main here: 

https://github.com/mrene/minidsp-rs/blob/54e6200d8bfbc0c62ad0f6d01c0d23bcbb3de809/devtools/src/main.rs#L165-L172

Then run: 
```
cargo run -p minidsp-devtools --no-default-features --features symbols -- codegen ./protocol/src/device/; cargo fmt
```

It'll create a file like this:
https://github.com/mrene/minidsp-rs/blob/54e6200d8bfbc0c62ad0f6d01c0d23bcbb3de809/protocol/src/device/m2x4hd.rs#L7-L12

Once the files are generated, add it to [probe.rs](https://github.com/mrene/minidsp-rs/blob/54e6200d8bfbc0c62ad0f6d01c0d23bcbb3de809/protocol/src/device/probe.rs) so it can recognize the device type (you can find the hardware id and dsp version with `minidsp probe`, this will make it recognize the device correctly)

Once this is done, you will be able to run `minidsp --force-kind ddrc88d -vvv server` and connect to it from the plugin (`--force-kind` isn't necessary if you configured the right ids inside `probe.rs`) . You will  see the actual symbols for each setting being modified (instead of its numerical value).. You then fiddle with every setting until you find their patterns and add them to `codegen/ddrc88d/mod.rs`. For the ones that I did, you can usually find the pattern for Input 1 and then replicate it for all inputs by changing some number in the symbol name.

See https://github.com/mrene/minidsp-rs/blob/54e6200d8bfbc0c62ad0f6d01c0d23bcbb3de809/devtools/src/codegen/m2x4hd/mod.rs#L22-L39

Then re-run codegen and try things out :) It will panic if you try to re-use a symbol twice, so if you can codegen the full thing you're in good shape.

Feel free to hit me up on discord (link in the docs) if you have questions.

**Reply by jvdh1978** [reply]
For me this was the first time looking at rs code, so that was a nice puzzle. I can't follow the code fully, but the things I do recognise seem very elegant. I matched the naming in the xml file in the RS file generation code and searched the project for all mentions of one of the devices to add an ddrc88bm entry. After some trial and error and recompilations, It seems to work now. The generated ddrc88bm.rs is missing new lines though so it looks somewhat messy. the vvv mode nicely seems to name the changed parameters I tried and I don't need to add the --force-kind option. I'm a self tough programmer so forgive me if I seem naive: How would you like me to share the result. I have all changes nicely tracked in GitHub desktop.

**Reply by mrene** [reply]
That's great! You can open a pull request.

**Reply by mrene** [reply]
Btw, if you run "cargo fmt", it will reformat the generated file so it looks more like normal code.
